### ISSUE TITLE
Labels for unlabeled data while training the model

### ISSUE THREAD

**Issue by padshahrohan** [ZeroRin/BertGCN#20]
As BertGCN and GCN in general is a good model for Semi supervised learning. I have a dataset which has a 2 class classification with lot of unlabeled data. In `finetune_bert.py` and `train_bert_gcn.py` labels are converted to class ID for pytorch computation. As unlabeled data is also passed to training dataset, what should be put in labels for this data. Going with current code it will give it a `0` class with the following code:

```
# transform one-hot label to class ID for pytorch computation
y = th.LongTensor((y_train + y_val +y_test).argmax(axis=1))
label = {}
label['train'], label['val'], label['test'] = y[:nb_train], y[nb_train:nb_train+nb_val], y[-nb_test:]
```

Can pytorch take one-hot labels as labels instead of class ID? Or Can any other approach be used?

**Comment by ZeroRin** [comment]
I suppose that marking only labeled data as train and others as test should do the trick.

**Comment by padshahrohan** [comment]
As per my understanding of semi-supervised model, we should be able to use unlabeled data to train the model, but i also know model would learn only from labeled data, so how will GCN use unlabeled data to train the model, I want to primarily do that

**Comment by ZeroRin** [comment]
the predict result of a data based on not only itself but also its neighbors (both labeled and unlabeled) so that unlabeled data will also be touched during training

**Comment by padshahrohan** [comment]
Does this mean that the input_ids and attention_mask will be created from entire data (labeled and unlabeled) with below line:
``` input_ids, attention_mask = encode_input(text, model.tokenizer) ```, the text variable will consist of all the data.

But ```label['train']``` will contain only labels of labeled training data, the indexes for unlabeled data will be skipped because the ```train_mask``` matrix will contain mask for only the training labels  (no val,test,unlabel).

Is my understanding correct that model will be trained on entire ```input_ids``` and the ```label['train']``` matrix is just used for calculating loss and accuracy?

**Comment by ZeroRin** [comment]
note that the embeddings of unlabeled data affects the predict result of labeled data during graph convolution, which means that bert will learn to extract embeddings from unlabeled data to help classifying labeled data.

**Comment by padshahrohan** [comment]
Can you just clarify one more question

Is my understanding correct that the ```label``` matrix is just used for calculating loss and accuracy?

**Comment by ZeroRin** [comment]
guess i'm not familiar with semi-supervised learning enough, i am a bit confused. are you trying to assign some kind of pseudo labels to unlabeled data and compute loss on them?

**Comment by padshahrohan** [comment]
I have total 8823 documents as follows:

200 labeled data
8623 unlabeled data

I want to use unlabeled data to train the model without assigning them any labels. so now how to use this in finetuning bert and training bert gcn?

**Comment by ZeroRin** [comment]
finetuning bert with unlabeled data is out of the scope of our work. our finetuning here mainly focus on learning towards the target classification task.

for training bert gcn, putting all unlabeled data in test set should do the trick. their embeddings will contribute to the classification of labeled nodes, and the bert model will learn to extract more helpful embeddings from unlabeled datas in the procedure.

**Comment by padshahrohan** [comment]
What should be the labels used when we put unlabeled data in test? because it will need some labels i suppose
